File: SlidingWindow/LongestSubstring.cpp
#include <vector>
#include <string>
#include <set>
#include <algorithm>

using namespace std;


class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        set<char> sSet;
        string subS = "";
        size_t maxLen = 0;
        for(int i = 0 ; i < s.length() ; i++)
        {
            if(sSet.find(s[i]) == sSet.end())
            {
                subS += s[i];
                sSet.insert(s[i]); 
            }
            else
            {
                while(sSet.find(s[i]) != sSet.end())
                {
                    sSet.erase(subS[0]);
                    subS.erase(0,1); 
                }
                subS += s[i];
                sSet.insert(s[i]); 
            }
            maxLen = max(maxLen, subS.length());
        }
        return maxLen;
    }
};

File: SlidingWindow/LongestSubstring.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        sSet = set()
        subS = ""
        maxLen = 0
        for i in range(len(s)):
            if s[i] in sSet:
                while s[i] in sSet:
                    sSet.remove(subS[0])
                    subS = subS[1:]
                sSet.add(s[i])
                subS += s[i]
            else:
                sSet.add(s[i])
                subS += s[i]
            maxLen = max(maxLen,len(sSet))
        return maxLen